File: JavaScript_/Dynamic Programming/palindromicSubstrings.js
const countSubstrings = (s) => {
  let count = 0;
  let matrix = [];

  for (let char of s) {
    matrix.push(Array(s.length).fill(0));
  }

  for (let i = 0; i < s.length; i++) {
    matrix[i][i] = 1;
    count++;
  }

  for (let col = 1; col < s.length; col++) {
    for (let row = 0; row < col; row++) {
      // this check palindrom for string length <=2
      if (row === col - 1 && s[col] === s[row]) {
        matrix[row][col] = 1;
        count++;
      }
      // this checks for string length 3 or more and inner string palindrom
      else if (matrix[row + 1][col - 1] === 1 && s[col] === s[row]) {
        matrix[row][col] = 1;
        count++;
      }
    }
  }
  return count;
};

console.log(countSubstrings("aaa"));

File: Python_/dp/palindromicSubstrings.py
"""
Given a string s, return the number of palindromic substrings in it.

A string is a palindrome when it reads the same backward as forward.

A substring is a contiguous sequence of characters within the string.

Input: s = "abc"
Output: 3
Explanation: Three palindromic strings: "a", "b", "c".

Input: s = "aaa"
Output: 6
Explanation: Six palindromic strings: "a", "a", "a", "aa", "aa", "aaa".
"""
s = "abc"

def countSubstrings(s):
    count = 0
    matrix = [[0 for _ in range(len(s))] for _ in range(len(s))]
    
    for i in range(len(s)):
        matrix[i][i] = 1
        count+=1

    for col in range(1, len(s)):
        for row in range(col):
            if row == col-1 and s[col] == s[row]:
                matrix[row][col] = 1
                count+=1
            elif matrix[row+1][col-1] == 1 and s[col]==s[row]:
                matrix[row][col]=1
                count+=1

    return count

print(
    countSubstrings(s)
)